src/Curl/CaseInsensitiveArray.php

```diff
@@ -2,72 +2,243 @@
+    /**
+     * Offset Get
+     *
+     * Return the stored data at the provided offset. The offset is converted to
+     * lowercase and the lookup is done on the Data store directly.
+     *
+     * @see https://secure.php.net/manual/en/arrayaccess.offsetget.php
+     *
+     * @param string $offset Offset to lookup.
+     *
+     * @return mixed The data stored at the offset.
+     *
+     * @access public
+     *
+     * @author Michael Mulligan <michael@bigroomstudios.com>
+     */
+    public function offsetGet($offset) {
+        $offsetlower = strtolower($offset);
+        return array_key_exists($offsetlower, $this->data) ? $this->data[$offsetlower] : NULL;
     }
 
-    public function count()
-    {
-        return count($this->container);
+    /**
+     * Count
+     *
+     * @see https://secure.php.net/manual/en/countable.count.php
+     *
+     * @param void
+     *
+     * @return int The number of elements stored in the Array.
+     *
+     * @access public
+     *
+     * @author Michael Mulligan <michael@bigroomstudios.com>
+     */
+    public function count() {
+        return (int) count($this->data);
     }
 
-    public function current()
-    {
-        return current($this->container);
+    /**
+     * Current
+     *
+     * @see https://secure.php.net/manual/en/iterator.current.php
+     *
+     * @param void
+     *
+     * @return mixed Data at the current position.
+     *
+     * @access public
+     *
+     * @author Michael Mulligan <michael@bigroomstudios.com>
+     */
+    public function current() {
+        return current($this->data);
     }
 
-    public function next()
-    {
-        return next($this->container);
+    /**
+     * Next
+     *
+     * @see https://secure.php.net/manual/en/iterator.next.php
+     *
+     * @param void
+     *
+     * @return void
+     *
+     * @access public
+     *
+     * @author Michael Mulligan <michael@bigroomstudios.com>
+     */
+    public function next() {
+        next($this->data);
     }
 
-    public function key()
-    {
-        return key($this->container);
+    /**
+     * Key
+     *
+     * @see https://secure.php.net/manual/en/iterator.key.php
+     *
+     * @param void
+     *
+     * @return mixed Case-Sensitive key at current position.
+     *
+     * @access public
+     *
+     * @author Michael Mulligan <michael@bigroomstudios.com>
+     */
+    public function key() {
+        $key = key($this->data);
+        return isset($this->keys[$key]) ? $this->keys[$key] : $key;
     }
 
-    public function valid()
-    {
-        return !($this->current() === false);
+    /**
+     * Valid
+     *
+     * @see https://secure.php.net/manual/en/iterator.valid.php
+     *
+     * @param void
+     *
+     * @return bool If the current position is valid.
+     *
+     * @access public
+     *
+     * @author Michael Mulligan <michael@bigroomstudios.com>
+     */
+    public function valid() {
+        return (bool) !(key($this->data) === NULL);
```

Comment:
Test the Key, as we know that the key will never equal `NULL` (this triggers a `push`) and using `current($array) !== FALSE` prevents the use of `bool` values (beyond our use-case, but this is a generic class, so let's do it right).